tetrominos_learning/tetrominos_learning.py

```diff
@@ -0,0 +1,291 @@
+
+# List of all the weights
+parameters = [
+    r1,
+    r2,
+    theta1,
+    phi1,
+    theta2,
+    phi2,
+    sqr1,
+    sqphi1,
+    sqr2,
+    sqphi2,
+    dr1,
+    dphi1,
+    dr2,
+    dphi2,
+    kappa1,
+    kappa2,
+]
+
+# ====================================================
+#   Definition of the quantum neural network
+# ====================================================
+
+# Single quantum variational layer
+
+
+def layer(l):
+    with tf.name_scope('layer_{}'.format(l)):
+        BSgate(theta1[l], phi1[l]) | (q[0], q[1])
+        Rgate(r1[l]) | q[0]
+        Sgate(tf.clip_by_value(sqr1[l], -sq_clip, sq_clip), sqphi1[l]) | q[0]
+        Sgate(tf.clip_by_value(sqr2[l], -sq_clip, sq_clip), sqphi2[l]) | q[1]
+        BSgate(theta2[l], phi2[l]) | (q[0], q[1])
+        Rgate(r2[l]) | q[0]
+        Dgate(tf.clip_by_value(dr1[l], -disp_clip, disp_clip), dphi1[l]) | q[0]
+        Dgate(tf.clip_by_value(dr2[l], -disp_clip, disp_clip), dphi2[l]) | q[1]
+        Kgate(tf.clip_by_value(kappa1[l], -kerr_clip, kerr_clip)) | q[0]
+        Kgate(tf.clip_by_value(kappa2[l], -kerr_clip, kerr_clip)) | q[1]
+
+
+# StrawberryFields quantum simulator
+engine, q = sf.Engine(num_subsystems=modes)
+
+# Definition of the CV quantum network
+with engine:
+    # State preparation
+    Dgate(disps_alpha) | q[0]
+    Dgate(disps_beta) | q[1]
+    # Sequence of variational layers
+    for i in range(depth):
+        layer(i)
+
+# Symbolic evaluation of the output state
+state = engine.run('tf', cutoff_dim=cutoff, eval=False, batch_size=num_images)
+ket = state.ket()
+
+# Projection on the subspace of up to im_dim-1 photons for each mode.
+ket_reduced = ket[:, :im_dim, :im_dim]
+norm = tf.sqrt(tf.abs(tf.reduce_sum(tf.conj(ket_reduced) * ket_reduced, axis=[1, 2])))
+norm_inv = tf.expand_dims(
+    tf.expand_dims(tf.cast(1.0 / norm, dtype=tf.complex64), axis=[1]), axis=[1]
+)
+ket_processed = ket_reduced * norm_inv
+
+# ====================================================
+# 		 Definition of the loss function
+# ====================================================
+
+# Target images
+data_states = tf.placeholder(tf.complex64, shape=[num_images, im_dim, im_dim])
+
+# Overlaps with target images
+overlaps = tf.abs(tf.reduce_sum(tf.conj(ket_processed) * data_states, axis=[1, 2])) ** 2
+
+# Overlap cost function
+overlap_cost = tf.reduce_mean((overlaps - 1) ** 2)
+
+# State norm cost function
+norm_cost = tf.reduce_sum(
+    (tf.abs(tf.reduce_sum(tf.conj(ket) * ket, axis=[1, 2])) ** 2 - 1) ** 2
+)
+
+cost = overlap_cost + norm_weight * norm_cost
+
+# ====================================================
+# 	TensorBoard logging of cost functions and images
+# ====================================================
+
+tf.summary.scalar('Cost', cost)
+tf.summary.scalar('Norm cost', norm_cost)
+tf.summary.scalar('Overlap cost', overlap_cost)
+
+# Output images with and without subspace projection.
+images_out = tf.abs(ket_processed)
```

Comment:
Should we have a squared here? E.g., `tf.abs(ket_processed) ** 2` so that we represent probabilities?

Reply:
Good point! Done.